### ISSUE TITLE
Normalize messages in the AS

### ISSUE THREAD

**Issue by willyaranda** [telefonicaid/notification_server#73]
Was: **_Do not allow replay messages: do not send signature back to the devices**_

And other things, we need to decide what parameters of the original JSON notification are passed to the device.

Right now, we have:

``` javascript
messageType,
id,
message,
signature,
ttl,
timestamp,
priority
```

`messageType` is mandatory (value `notification`), `message` is the payload. And maybe the `id` could be used to replace data in the final app.

@frsela ?

**Comment by frsela** [comment]
Well, from my point of view, to the UA we only should sent:

{
  messageType: "notification",
  id: <id>,  // To be used on the ACK ;)
  message: <payload>,
  timestamp: <epoch>   // Not sure if needed ... anyway can be used by the client to know how old the message is. Anyway, if the app needs this can also add it to the message payload ... WDYT?
}

**Comment by jmcanterafonseca** [comment]
do we really need messageType: notification, can we consider that the default is a notification and save those bytes? 

I think an automatic timestamp is useful and helps apps

**Comment by frsela** [comment]
Well, we can avoid messageType to reduce the amount of data to be sent but in order to have a consistent protocol I think is better to maintain it. Only to be more clear ;)

I think we should have a "low-bandwidth" version of the protocol reducing the length of the literals like:

{
  y: "n",
  i: "id",
  m: "message",
  t: "timestamp"
}

Also we should check if HTTP gzip transparent compression is being used in the node.js server which also reduces payload size ;)

**Comment by frsela** [comment]
@willyaranda on the API doc. we shared with mozilla, we also include the URL (used by thinker to identify the app) and the messageId (our internal id):

{
    "messageType" : "notification" ,
    "id" : 1234 ,
    "message" : "Message" , 
    "signature" : "MessageSigned" , 
    "ttl" : 0 , 
    "timestamp" : "SINCE_EPOCH_TIME" , 
    "priority" : 1 , 
    "messageId" : "dc3eab10-eddf-11e1-b4ab-c7f2c558356e" , 
    "url" :     "http://localhost:8081/notify/b578c949551625b05bcac73fb81a9d8bf9272bb0c2a47ff8f7bc6b8bb44554ef"
}

The final message can be:

{
    "messageType" : "notification" ,
    "id" : "dc3eab10-eddf-11e1-b4ab-c7f2c558356e" ,
    "payload" : {
                id: "123456",                           // 3rd. party message id
                message: "Message"
        } , 
    "timestamp" : "SINCE_EPOCH_TIME" , 
    "url" : "http://localhost:8081/notify/b578c949551625b05bcac73fb81a9d8bf9272bb0c2a47ff8f7bc6b8bb44554ef"
}

WDYT? I'll update the doc, but I cann't update the wiki since I haven't write permissions on it.

**Comment by willyaranda** [comment]
Yes, we can avoid the "messageType" in the POST with notification, we just use that to be consistent, as @frsela said. But we can delete it.

BTW, when you POST to a URL, you can post it gzipped? I think it's only the responses, right?

PS: just create a user in the wiki, it's writable by anyone with an account.

**Comment by willyaranda** [comment]
In fact… someone could trick our system :) Because we just allow 1024 bytes in the message payload, but without limit in others. We do not check if there are more non-standard attributes in the JSON and just save the notification in the system, sending back the notifications to the user… with those non-standard fields. We need to normalize our notifications in the server.

@frsela this is interesting ;)

**Comment by frsela** [comment]
Yes, this is also very important.

Good raised point :)

Enviado de Samsung Mobile

Guillermo López Leal notifications@github.com escribió:

In fact… someone could trick our system :) Because we just allow 1024 bytes in the message payload, but without limit in others. We do not check if there are more non-standard attributes in the JSON and just save the notification in the system, sending back the notifications to the user… with those non-standard fields. We need to normalize our notifications in the server.

@frselahttps://github.com/frsela this is interesting ;)

—
Reply to this email directly or view it on GitHubhttps://github.com/telefonicaid/notification_server/issues/73#issuecomment-8728076.

---

Este mensaje se dirige exclusivamente a su destinatario. Puede consultar nuestra política de envío y recepción de correo electrónico en el enlace situado más abajo.
This message is intended exclusively for its addressee. We only send and receive email on the basis of the terms set out at:
http://www.tid.es/ES/PAGINAS/disclaimer.aspx

**Comment by willyaranda** [comment]
Fixed in commit 9476eaffa1da4f9c770da7be73e261d85753e3c3
